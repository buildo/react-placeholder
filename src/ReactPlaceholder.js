import React from 'react';
import PropTypes from 'prop-types';
import * as placeholders from './placeholders';

export default class ReactPlaceholder extends React.Component {

  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.element
    ]).isRequired,
    delay: PropTypes.number,
    ready: PropTypes.bool.isRequired,
    firstLaunchOnly: PropTypes.bool,
    type: PropTypes.oneOf(['text', 'media', 'textRow', 'rect', 'round']),
    rows: PropTypes.number,
    color: PropTypes.string,
    showLoadingAnimation: PropTypes.bool,
    customPlaceholder: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.element
    ]),
    holdPlaceholder: PropTypes.number
  }

  static defaultProps = {
    delay: 0,
    type: 'text',
    color: '#CDCDCD',
    holdPlaceholder: 0
  }

  state = {
    ready: this.props.ready,
    holdPlaceholder: this.props.holdPlaceholder > 0
  }

  getFiller = () => {
    const {
      firstLaunchOnly, children, ready, className, holdPlaceholder, // eslint-disable-line no-unused-vars
      type, customPlaceholder, showLoadingAnimation, ...rest
    } = this.props;

    const classes = showLoadingAnimation ?
      ['show-loading-animation', className].filter(c => c).join(' ') :
      className;

    if (customPlaceholder && React.isValidElement(customPlaceholder)) {
      const mergedCustomClasses = [
        customPlaceholder.props.className,
        classes
      ].filter(c => c).join(' ');
      return React.cloneElement(customPlaceholder, { className: mergedCustomClasses });
    } else if (customPlaceholder) {
      return customPlaceholder;
    }

    const Placeholder = placeholders[type];

    return <Placeholder {...rest} className={classes} />;
  };

  setNotReady = () => {
    const { delay } = this.props;

    if (delay > 0) {
      this.timeout = setTimeout(() => {
        this.setState({ ready: false });
      }, delay);
    } else {
      this.setState({ ready: false });
    }
  }

  setReady = () => {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }

    if (!this.state.ready) {
      this.setState({ ready: true });
    }
  }

  render() {
    const { ready, holdPlaceholder } = this.state;
    return ready && !holdPlaceholder ? this.props.children : this.getFiller();
  }

  componentWillReceiveProps(nextProps) {
    if (!this.props.firstLaunchOnly && !nextProps.ready) {
      if (this.state.ready) {
        this.setNotReady();
      }
      if (this.props.holdPlaceholder) {
        clearTimeout(this.holdPlaceholderTimeout);
        this.setState({ holdPlaceholder: true });
        this.holdPlaceholderTimeout = setTimeout(() => this.setState({ holdPlaceholder: false }), nextProps.holdPlaceholder);  
      }
    } else if (nextProps.ready) {
      this.setReady();
    }
  }

  componentWillMount() {
    if (this.props.holdPlaceholder > 0) {
      this.holdPlaceholderTimeout = setTimeout(() => this.setState({ holdPlaceholder: false }), this.props.holdPlaceholder);
    }
  }
}
