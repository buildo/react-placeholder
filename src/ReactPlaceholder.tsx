import * as React from 'react';
import * as PropTypes from 'prop-types';
import * as placeholders from './placeholders';

export type CommonProps = {
  children: React.ReactNode,
  /** pass `true` when the content is ready and `false` when it's loading */
  ready: boolean,
  /** delay in millis to wait when passing from ready to NOT ready */
  delay?: number,
  /** if true, the placeholder will never be rendered again once ready becomes true, even if it becomes false again */
  firstLaunchOnly?: boolean,
  className?: string,
  style?: React.CSSProperties
}

export type Props = (CommonProps & {
  /** type of placeholder to use */
  type: 'text' | 'media' | 'textRow' | 'rect' | 'round',
  /** number of rows displayed in 'media' and 'text' placeholders */
  rows?: number,
  /** color of the placeholder */
  color?: string,
  /** pass true to show a nice loading animation on the placeholder */
  showLoadingAnimation?: boolean,
  customPlaceholder?: undefined
}) | (CommonProps & {
  /** pass any renderable content to be used as placeholder instead of the built-in ones */
  customPlaceholder?: React.ReactNode | React.ReactElement<{ [k: string]: any }>,
  type?: undefined,
  rows?: undefined,
  color?: undefined,
  showLoadingAnimation?: undefined
})

export default class ReactPlaceholder extends React.Component<Props> {

  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.element
    ]).isRequired,
    delay: PropTypes.number,
    ready: PropTypes.bool.isRequired,
    firstLaunchOnly: PropTypes.bool,
    type: PropTypes.oneOf(['text', 'media', 'textRow', 'rect', 'round']),
    rows: PropTypes.number,
    color: PropTypes.string,
    showLoadingAnimation: PropTypes.bool,
    customPlaceholder: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.element
    ]),
    className: PropTypes.string,
    style: PropTypes.object
  }

  static defaultProps = {
    delay: 0,
    type: 'text',
    color: '#CDCDCD'
  }

  state = {
    ready: this.props.ready
  }

  timeout?: number;

  getFiller = () => {
    const {
      firstLaunchOnly, children, ready, className, // eslint-disable-line no-unused-vars
      type, customPlaceholder, showLoadingAnimation, ...rest
    } = this.props;

    const classes = showLoadingAnimation ?
      ['show-loading-animation', className].filter(c => c).join(' ') :
      className;

    if (customPlaceholder && React.isValidElement(customPlaceholder)) {
      const mergedCustomClasses = [
        customPlaceholder.props.className,
        classes
      ].filter(c => c).join(' ');
      return React.cloneElement(customPlaceholder, { className: mergedCustomClasses });
    } else if (customPlaceholder) {
      return customPlaceholder;
    }

    const Placeholder = placeholders[type!];

    return <Placeholder {...rest} className={classes} />;
  };

  setNotReady = () => {
    const { delay } = this.props;

    if (delay && delay > 0) {
      this.timeout = window.setTimeout(() => {
        if (!this.props.ready) {
          this.setState({ ready: false });
        }
      }, delay);
    } else {
      this.setState({ ready: false });
    }
  }

  setReady = () => {
    if (this.timeout) {
      window.clearTimeout(this.timeout);
    }

    if (!this.state.ready) {
      this.setState({ ready: true });
    }
  }

  render() {
    return this.state.ready ? this.props.children : this.getFiller();
  }

  componentWillReceiveProps(nextProps: Props) {
    if (!this.props.firstLaunchOnly && this.state.ready && !nextProps.ready) {
      this.setNotReady();
    } else if (nextProps.ready) {
      this.setReady();
    }
  }
}
