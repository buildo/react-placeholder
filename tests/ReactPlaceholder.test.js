import React from 'react';
import ReactPlaceholder from '../src/ReactPlaceholder';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

jest.useFakeTimers();

describe('ReactPlaceholder', () => {

  it('renders the placeholder when the content is not ready', () => {
    const content = <div>Some content still loading...</div>;
    const tree = shallow(
      <ReactPlaceholder type='text' rows={2} ready={false}>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the placeholder only after the specified delay', () => {
    const content = <div>Some content still loading...</div>;
    const tree = shallow(
      <ReactPlaceholder
        ready
        type='text'
        rows={2}
        delay={10000}
      >
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(true);
    tree.setProps({ ready: false });
    tree.update();
    expect(tree.contains(content)).toBe(true);
    jest.runAllTimers();
    tree.update();
    expect(tree.contains(content)).toBe(false);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the content when it\'s ready', () => {
    const content = <div>Some ready content</div>;
    const tree = shallow(
      <ReactPlaceholder type='text' rows={2} ready>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(true);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the placeholder for a while even if the content is ready', () => {
    const content = <div>Some ready content</div>;
    const tree = shallow(
      <ReactPlaceholder
        ready
        type='text'
        rows={2}
        holdPlaceholder={10000}
      >
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    jest.runAllTimers();
    tree.update();
    expect(tree.contains(content)).toBe(true);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the placeholder again for a while even if the content is ready', () => {
    const content = <div>Some ready content</div>;
    const tree = shallow(
      <ReactPlaceholder
        ready
        type='text'
        rows={2}
        holdPlaceholder={10000}
      >
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    jest.runAllTimers();
    tree.update();
    expect(tree.contains(content)).toBe(true);
    tree.setProps({ ready: false });
    tree.update();
    expect(tree.contains(content)).toBe(false);
    tree.setProps({ ready: true });
    tree.update();
    expect(tree.contains(content)).toBe(false);
    jest.runAllTimers();
    tree.update();
    expect(tree.contains(content)).toBe(true);
    expect(tree.getElements()).toMatchSnapshot();
  });

});
