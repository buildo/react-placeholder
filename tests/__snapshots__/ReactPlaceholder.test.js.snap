// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactPlaceholder renders the content when it's ready 1`] = `
Array [
  <div>
    Some ready content
  </div>,
]
`;

exports[`ReactPlaceholder renders the placeholder again for a while even if the content is ready 1`] = `
Array [
  <div>
    Some ready content
  </div>,
]
`;

exports[`ReactPlaceholder renders the placeholder for a while even if the content is ready 1`] = `
Array [
  <div>
    Some ready content
  </div>,
]
`;

exports[`ReactPlaceholder renders the placeholder only after the specified delay 1`] = `
Array [
  <TextBlock
    className={undefined}
    color="#CDCDCD"
    delay={10000}
    rows={2}
  />,
]
`;

exports[`ReactPlaceholder renders the placeholder when the content is not ready 1`] = `
Array [
  <TextBlock
    className={undefined}
    color="#CDCDCD"
    delay={0}
    rows={2}
  />,
]
`;
